Add animation reset track feature

RandomShaper opened the pull request:
This implements https://github.com/godotengine/godot-proposals/issues/1597.

**NOTE:** This is barely tested, so testing is welcome.

A few notes follow, about aspects not included or not settled on the original proposal:

---

~~Any animation can be the reset animation.~~

~~_Toggle Reset_ is used to make the current animation the reset one, or revert that:~~

~~When clicked, you get one of these:~~

~~The option to toggle reset is there and also asks for confirmation, to avoid accidental clicks that could happen if it was just a button in the animation dock. Also, there's no frequent need to deal with it.~~

---

The reset animation is marked as such in the dropdown:

![image](https://user-images.githubusercontent.com/11797174/97243891-92ab6880-17f7-11eb-906c-63177e09b0a4.png)

If it's the autoplay animation at the same time, it looks like this:

![image](https://user-images.githubusercontent.com/11797174/97243837-7a3b4e00-17f7-11eb-81ea-06975f5ced2b.png)

---

If you have _Create Reset Track(s)_ checked, and there's no reset animation, one will be created with the name _RESET_.

---

You can apply the reset animation any time, unless the it's the current one:

![image](https://user-images.githubusercontent.com/11797174/97244135-1feebd00-17f8-11eb-841c-f08629dbc167.png)

---

There's a new `reset_on_save` property:

![image](https://user-images.githubusercontent.com/11797174/97244535-46612800-17f9-11eb-87f9-9dca1049e900.png)

The cool thing is that it helps in getting a cleaner diff and consistent saved scene, without breaking the workflow, since the current state in the editor is kept. If you need to see how its saved version looks, you can use _Scene-> Reload Saved Scene_.

RandomShaper committed:
Extend UndoRedo handling of Resource to every Reference

reduz commented:
My first feedback is that you should never create the reset track manually, or at much I would add an option in the New dialog like "Create RESET track" and make sure the name is always this (can't be something else and can't be renamed). This is unnecessary flexibility and I think users will appreciate more that this track as a standard name.

![image](https://user-images.githubusercontent.com/6265307/97247161-dbeebe80-17dd-11eb-8a05-5d52514acb03.png)

Of course option will be grayed out if the track already exists.

That said, my belief is that the common way to create this track would be to never do it manually. Instead, it should be a single checkbox in the new track create dialog:

![image](https://user-images.githubusercontent.com/6265307/97247429-6df6c700-17de-11eb-820b-2302ccdcb415.png)

This checkbox exists by default here and is turned on by default. If you don't want to use it, you toggle it off in the project settings.

![image](https://user-images.githubusercontent.com/6265307/97247748-08570a80-17df-11eb-8450-adf1755cc2bd.png)

This new setting also takes effect if you disable confirmation in this same section (when no confirmation is active and setting is enabled, RESET is track is always created).

DDL-Blue commented:
Is it even a good idea to have a reset ANIMATION? Don't we want just reset KEYFRAME? I know this might fit the UI well, but it make no sense to me to be an animation. Or do I miss something?

fire commented:
Shouldn't this be a import setting? For like FBX, GLTF etc.

reduz commented:
@DDL-Blue This has been discussed in the original proposal. Besides it being easier to implement this way and offer a familiar way to edit it, it allows to queue it together with other animations (play something, then queue something, then queue reset, then another). Users have already been doing this workflow manually in their project, so the main goal of this approach is to just formalize it and make it easier to use and discover.

reduz commented:
@fire Yes sure, would be nice to implement it once this is merged.

DDL-Blue commented:
@reduz So when you save, it will (if enabled) play the whole animation? In some fast-forward way? Trigger all the function calls? Hopefully not play the sounds. I got your points, but this still seems a bit hacky to me.

RandomShaper commented:
@reduz, even with a fixed name for the reset animation (so you can't set any animation as reset), you'll still be able to insert reset tracks manually by editing the reset animation, because in the end it's just an animation.

I'm asking to clarify if that's what you want.

RandomShaper committed:
Add animation reset track feature

As a bonus, to have consistency between use Beziers and create insert tracks, use Beziers also gets a default via editor settings that is used when the confirmation dialog is disabled, instead of just falling back to creating non-Bezier tracks.

RandomShaper commented:
Updated with the requested changes.

One of the more important changes is this:
![image](https://user-images.githubusercontent.com/11797174/102712493-0df23d80-42c2-11eb-96d0-abb954a01ded.png)
Both defaults will inform the initial checked state of their corresponding check boxes in the insert tracks dialog (one time. at the start of the editor session) and also will be used when the confirmation dialog is disabled. This keeps both insert track settings consistent, while also making it possible now to insert Bezier tracks when the confirmation dialog is disabled.

RandomShaper commented:
> So when you save, it will (if enabled) play the whole animation? In some fast-forward way? Trigger all the function calls?

It will play every key at time 0 in the RESET animation (the description of the setting has been updated to clarify that).

akien-mga commented:
Thanks!

Sslaxx commented:
Fixes #812?

RandomShaper commented:
> Fixes #812?

I think so. I see @reduz already put there what we could call the first draft of the proposal. So, yes, this is finally the incarnation of that (if I've understood that issue correctly).
